service/tablet_allocator.cc

```diff
@@ -649,10 +676,60 @@ class load_balancer {
         return utils::get_local_injector().enter("tablet_allocator_shuffle");
     }
 
-    shard_id rand_shard(shard_id shard_count) const {
+    size_t rand_int() const {
         static thread_local std::default_random_engine re{std::random_device{}()};
-        static thread_local std::uniform_int_distribution<shard_id> dist;
-        return dist(re) % shard_count;
+        static thread_local std::uniform_int_distribution<size_t> dist;
+        return dist(re);
+    }
+
+    shard_id rand_shard(shard_id shard_count) const {
+        return rand_int() % shard_count;
+    }
+
+    table_id pick_table(const std::unordered_map<table_id, std::unordered_set<global_tablet_id>>& candidates) {
+        if (!_use_table_aware_balancing) {
+            on_internal_error(lblogger, "pick_table() called when table-aware balancing is disabled");
+        }
+        size_t total = 0;
+        for (auto&& [table, tablets] : candidates) {
+            total += tablets.size();
+        }
+        ssize_t candidate_index = rand_int() % total;
+        for (auto&& [table, tablets] : candidates) {
+            candidate_index -= tablets.size();
+            if (candidate_index <= 0 && !tablets.empty()) {
+                return table;
+            }
```

Comment:
All these loops on table count probably won't stall, but do look wasteful.

Reply:
Isn't picking a random global tablet the same as this?

Reply:
We need to pick from candidates, which is not an indexable container. We could pick global tablet at random and check if it's a candidate, but there's no upper bound on such iterations.

Reply:
There's candidates_all_tables and there's candidates. Isn't picking a random item from candidates_all_tables equivalent to picking a random table (weighted by the number of candidates) and then picking a random tablet from the table?

Reply:
Yes, but candidates_all_tables is an unordered_set so we cannot cheaply index into it.